lib/bolt/fiber_executor.rb

```diff
@@ -0,0 +1,168 @@
+          block.closure.call_by_name_with_scope(newscope, args, true)
+        end
+
+        # If we got a return from the block, get it's value
+        # Otherwise the result is the last line from the block
+        result = result.value if result.is_a?(Puppet::Pops::Evaluator::Return)
+
+        # Validate the result is a PlanResult
+        unless Puppet::Pops::Types::TypeParser.singleton.parse('Boltlib::PlanResult').instance?(result)
+          raise Bolt::InvalidParallelResult.new(result.to_s, *Puppet::Pops::PuppetStack.top_of_stack)
+        end
+
+        result
+      rescue Puppet::PreformattedError => e
+        if e.cause.is_a?(Bolt::Error)
+          e.cause
+        else
+          raise e
+        end
+      end
+
+      # PlanFutures are assigned an ID, which is just a global incrementing
+      # integer. The main plan should always have ID 0.
+      @id += 1
+      Bolt::PlanFuture.new(future, @id, name)
+    end
+
+    # Register a PlanFuture with the FiberExecutor to be executed while round_robining.
+    def add_future(future)
+      # This is a little incongruous since we created the future in
+      # `create_future`, but a future can't do anything until it's added to
+      # plan_futures, and all futures (main plan and fiber) flow through the
+      # method.
+      msg = "Created future #{future.name}"
+      @logger.trace(msg)
+
+      plan_futures << future
+    end
+
+    # Visit each PlanFuture registered with the FiberExecutor and resume it.
+    # Fibers will yield themselves back, either if they kicked off a
+    # long-running process or if the current long-running process hasn't
+    # completed. If the Fiber finishes after being resumed, store the result in
+    # the PlanFuture and remove the PlanFuture from the FiberExecutor.
+    def round_robin
+      plan_futures.each do |future|
+        # If the Fiber is still running and can be resumed, then resume it
+        @logger.trace("Checking future '#{future.name}'")
+        if future.alive?
+          @logger.trace("Resuming future '#{future.name}'")
+          future.resume
+        end
+
+        # Once we've restarted the Fiber, check to see if it's finished again
+        # and cleanup if it has.
+        next if future.alive?
+        @logger.trace("Cleaning up future '#{future.name}'")
+
+        # If the future errored and the main plan has already exited, log the
+        # error at warn level.
+        unless plan_futures.map(&:id).include?(0) || future.state == "done"
+          Bolt::Logger.warn('errored-futures', "Error in future '#{future.name}': #{future.value}")
+        end
+
+        # Remove the PlanFuture from the FiberExecutor.
+        plan_futures.delete(future)
+      end
+
+      # If the Fiber immediately returned wait for long enough that something
+      # can execute before checking again. This mitigates CPU thrashing.
+      return unless plan_futures.all? { |f| f.value.is_a?(Symbol) && f.value == :returned_immediately }
```

Comment:
```suggestion
      return unless plan_futures.all? { |f| f.value == :returned_immediately }
```

I'm not sure that the class check is needed, since the equality operator should do the same?